### PR TITLE
Add race testing to tests, and fix a few small race conditions in the tests

### PR DESCRIPTION
**Pull request by brianathere**
### Pull Request Checklist

In our CI pipeline the unit tests are non-deterministically failing. The first step in addressing this I propose is to make sure they don't have race conditions that are detected by the race checker.

* [X] I have added added tests for PR _or_ I have justified why this PR doesn't need tests.
* [X] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/main/docs/CONTRIBUTING.md#sign-off)

Signed-off-by: `Brian Meek <brian@hntlabs.com>`

### CODE REVIEW

**Comment by neilalexander** [comment]
Can you please update your sign-off line?

**Review by neilalexander — CHANGES_REQUESTED** [review]
Although I do agree that the mutexes are good to have here for avoiding races in these tests, I'm not mad on bringing in more dependencies (like `github.com/sasha-s/go-deadlock`) to achieve it.

File: syncapi/sync/requestpool_test.go [hunk 1]
```diff
@@ -125,11 +128,19 @@ func TestRequestPool_updatePresence(t *testing.T) {
 	go rp.cleanPresence(db, time.Millisecond*50)
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			beforeCount := publisher.count
+			beforeCount := func() int {
```
Comment (S7evinK) [inline syncapi/sync/requestpool_test.go:131]: Probably easier to `Lock`/`Unlock` before/after setting `beforeCount`, for readability.

File: internal/sqlutil/migrate.go [hunk 2]
```diff
@@ -50,6 +50,7 @@ type Migration struct {
 
 // Migrator
 type Migrator struct {
+	gooseMutex      sync.Mutex
```
Comment (S7evinK) [inline internal/sqlutil/migrate.go:53]: Do you see any races without this mutex?
There's already a mutex on this struct, which is protecting `knownMigrations`, the other maps shouldn't be accessed concurrently.

Comment (brianathere) [reply internal/sqlutil/migrate.go:53]: Yes, the race is coming from accessing goose in two separate goroutines, I'll add a comment clarifying the need for this lock.

Comment (S7evinK) [reply internal/sqlutil/migrate.go]: Oh, I just saw that this PR was created before #2264 was merged. So I _think_ we may not need this mutex at all.

Comment (brianathere) [reply internal/sqlutil/migrate.go]: Ah, nice, thank you, I missed this. I removed the mutex.

**Review by neilalexander — APPROVED** [review]

**Pull request merged**